src/iterative_ensemble_smoother/esmda.py

```diff
@@ -0,0 +1,299 @@
+        ensemble_mask: Optional[npt.NDArray[np.bool_]] = None,
+    ) -> npt.NDArray[np.double]:
+        """Assimilate data and return an updated ensemble.
+
+        Parameters
+        ----------
+        X : np.ndarray
+            2D array of shape (num_inputs, num_ensemble_members).
+        Y : np.ndarray
+            2D array of shape (num_ouputs, num_ensemble_members).
+        ensemble_mask : np.ndarray
+            1D boolean array of length `num_ensemble_members`, describing which
+            ensemble members are active. Inactive realizations are ignored.
+            Defaults to all active.
+
+        Returns
+        -------
+        X_posterior : np.ndarray
+            2D array of shape (num_inputs, num_ensemble_members).
+
+        """
+        if self.iteration >= self.num_assimilations():
+            raise Exception("No more assimilation steps to run.")
+
+        # Verify shapes
+        num_inputs, num_ensemble = X.shape
+        num_outputs, num_emsemble2 = Y.shape
+        assert (
+            num_ensemble == num_emsemble2
+        ), "Number of ensemble members in X and Y must match"
+        assert (ensemble_mask is None) or (
+            ensemble_mask.ndim == 1 and len(ensemble_mask) == num_ensemble
+        )
+
+        if ensemble_mask is None:
+            ensemble_mask = np.ones(num_ensemble, dtype=bool)
+
+        # Sample from a zero-centered multivariate normal with cov=C_D
+        mv_normal_rvs = self.mv_normal.rvs(
+            size=ensemble_mask.sum(), random_state=self.rng
+        ).T
+
+        # Line 2 (b) in the description of ES-MDA in the 2013 Emerick paper
+        # Create perturbed observationservations, with C_D scaled by alpha
+        # If C_D = L L.T  by the cholesky factorization, then
+        # drawing from a zero cented normal is y := L @ z, where z ~ norm(0, 1)
+        # Scaling C_D by alpha is equivalent to scaling L with sqrt(alpha)
+        D = (
+            self.observations[:, ensemble_mask]
+            + np.sqrt(self.alpha[self.iteration]) * mv_normal_rvs
+        )
+
+        # Line 2 (c) in the description of ES-MDA in the 2013 Emerick paper
+        # Compute the cross covariance
+        C_MD = empirical_cross_covariance(X[:, ensemble_mask], Y[:, ensemble_mask])
+
+        # Choose inversion method, e.g. 'exact'
+        inversion_func = self._inversion_methods[self.inversion]
+        K = inversion_func(
+            alpha=self.alpha[self.iteration], C_D=self.C_D, D=D, Y=Y[:, ensemble_mask]
+        )
+
+        # X_posterior = X_current + C_MD @ K
+        # K := sp.linalg.inv(C_DD + C_D_alpha) @ (D - Y)
+
+        # In the typical case where num_outputs >> num_inputs >> ensemble members,
+        # multiplying in the order below from the right to the left, i.e.,
+        # C_MD @ (inv(C_DD + alpha * C_D) @ (D - Y))
+        # is faster than the alternative order
+        # (C_MD @ inv(C_DD + alpha * C_D)) @ (D - Y)
+        X_posterior = np.copy(X)
```

Comment:
We should aim for an implementation that does not make a copy of X as it can be large.

Reply:
good point!